docket/common/utils.py

```diff
@@ -1,5 +1,252 @@
+        _overwrites_ existing values even if the new value is None
+        you might think deepcopy for this, but it can't merge two dictionaries
+    """
+    for k,v in b.items():
+        if isinstance(v, Mapping):
+            a[k] = recurse_update(a.get(k, {}), v)
+        else:
+            a[k] = v
+    return a
+
+def file_lock(f, timeout=2):
+    """ True if f was locked, False otherwise """
+    giveUp = datetime.utcnow() + timedelta(seconds=timeout)
+    while datetime.utcnow() < giveUp:
+        try:
+            flock(f, LOCK_EX | LOCK_NB)
+            return True
+        except IOError as e:
+            if e.errno == errno.EAGAIN:
+                time.sleep(0.1)
+    return False
+
+def update_yaml(path, val = None):
+    """ given a path to a yaml file:
+            if val: update val from the file, update the file, and return val.
+            else:   load the file and return that.
+        returns
+            False - locking failed, no update occurred
+            new value - File data, updated with provided value
+    """
+    if not path:
+        return False
+    locked = False
+
+    with open(path, 'a+b') as f:
+        if not file_lock(f, timeout=Config.get("LOCK_TIMEOUT", 2)):
+            Config.logger.error("Failed to lock {}".format(path))
+            return False
+
+        f.seek(0)
+        data = f.read()
+        if len(data):
+            tmp = yaml.load(data)
+            if not val:
+                Config.logger.debug("loaded {}".format(path))
+                # No changes to write.
+                return tmp
+            elif hasattr(val, 'update'):
+                Config.logger.debug("{}.update( {} )".format(val,tmp))
+                val.update(tmp)
+            elif hasattr(val, '__add__'):
+                Config.logger.debug("{} += {}".format(val,tmp))
+                val += tmp
+            else:
+                Config.logger.error("loaded {}, can't update {}".format(path,val))
+                return val
+
+        f.seek(0)
+        f.truncate()
+        f.write(yaml.dump(val))
+    return val
+
+def free_space(f):
+    """ return ( Bytes Free, Nodes free ) for the filesystem containing the provided file
+        ex: free_space( open( '/tmp/nonexistant', 'w' ) ) -> ( bytes=123456789 , nodes=1234 )
+    """
+    vfs = os.fstatvfs(f.fileno())
+    return FreeSpace(bytes = vfs.f_bavail * vfs.f_frsize, nodes = vfs.f_favail)
+
+def spool_space():
+    """ return free space in SPOOL_DIR """
+    fn = os.path.join( Config.get('SPOOL_DIR'), 'docket.yaml' )
+    with open(fn, 'a+b') as f:
+        return free_space(f)
+
+def space_low():
+    space = spool_space()
+    if space.bytes < parse_capacity(Config.get('FREE_BYTES', '100MB')):
+        return "Low on disk space: {} free".format(space.bytes)
+    if space.nodes < Config.get('FREE_NODES', 100 ):
+        return "Low on FileSystem Nodes: {} free".format(space.nodes)
+        return False
+
+def validate_ip(ip):
```

Comment:
You should really swap out these IP parsing functions for just using the [netaddr](https://pypi.python.org/pypi/netaddr) library. It's pretty prevalent and does all the things you ever wanted.

I didn't do this validation before because I just let stenographer handle it. It's probably _nicer_ to do it here to help the user.

Reply:
Since I'm queuing requests, I can't immediately provide the stenographer response to the user.  I believe fast feedback is extremely important for a user interface.

validate_ip() is just a little helper function around the standard library's presentation to network function. I investigated several options (regex - bad idea, 3rd party library, etc) and this seemed to provide all needed functionality without increasing requirements. 

If there's a downside to inet_pton(), I'm not aware of it.